src/libs/actions/Report.ts

```diff
@@ -2515,20 +2578,98 @@ function inviteToRoom(reportID: string, inviteeEmailsToAccountIDs: Record<string
+                    revertedParticipants[accountID] = null;
+                    return revertedParticipants;
+                }, {}),
                 pendingChatMembers: report?.pendingChatMembers ?? null,
             },
         },
         ...newPersonalDetailsOnyxData.finallyData,
     ];
 
+    if (ReportUtils.isGroupChat(report)) {
+        const parameters: InviteToGroupChatParams = {
+            reportID,
+            inviteeEmails,
+            accountIDList: newAccountIDs.join(),
+        };
+
+        API.write(WRITE_COMMANDS.INVITE_TO_GROUP_CHAT, parameters, {optimisticData, successData, failureData});
+        return;
+    }
+
     const parameters: InviteToRoomParams = {
         reportID,
         inviteeEmails,
     };
 
+    // eslint-disable-next-line rulesdir/no-multiple-api-calls
     API.write(WRITE_COMMANDS.INVITE_TO_ROOM, parameters, {optimisticData, successData, failureData});
 }
 
+function updateGroupChatMemberRoles(reportID: string, accountIDList: number[], role: ValueOf<typeof CONST.REPORT.ROLE>) {
+    const participants: Participants = {};
+    const memberRoles: Record<number, string> = {};
+    accountIDList.forEach((accountID) => {
+        memberRoles[accountID] = role;
+        participants[accountID] = {role};
+    });
+
+    const optimisticData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+            value: {
+                participants,
+            },
+        },
+    ];
+
+    API.write(WRITE_COMMANDS.UPDATE_GROUP_CHAT_MEMBER_ROLES, {reportID, memberRoles: JSON.stringify(memberRoles)}, {optimisticData});
+}
+
+function removeFromGroupChat(reportID: string, accountIDList: number[]) {
+    const removeParticipantsData: Record<number, null> = {};
+    const currentParticipants = ReportUtils.getParticipants(reportID);
+    const participantAccountIDs = [];
+    const visibleChatMemberAccountIDs = [];
+    for (const accountIDKey in currentParticipants) {
+        if (Object.hasOwn(currentParticipants, accountIDKey)) {
+            const participantAccountID = Number(accountIDKey);
+            const participant = currentParticipants[participantAccountID];
+            if (!accountIDList.includes(participantAccountID)) {
+                participantAccountIDs.push(participantAccountID);
+                if (!participant.hidden) {
+                    visibleChatMemberAccountIDs.push(participantAccountID);
+                }
+            }
+        }
+    }
+
+    accountIDList.forEach((accountID) => {
+        removeParticipantsData[accountID] = null;
+    });
+    const optimisticData: OnyxUpdate[] = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
+            value: {
+                participants: removeParticipantsData,
+                participantAccountIDs,
+                visibleChatMemberAccountIDs,
+            },
+        },
+    ];
```

Comment:
There is not any specific error handling in the backend for this flow. Not much could go wrong besides incorrect permissions or incorrect chat type. Both would be indicative of a problem in the front end code and not really something a user could do anything about.

That said, I am seeing some pretty involved work around removing members from rooms [here](https://github.com/Expensify/App/blob/05eb481875faaafa1bcd2077a3c4d10a36a89133/src/libs/actions/Report.ts#L2535-L2582) which seems to support the "Offline" case.

We have a concept of `pendingChatMembers` that I am still wrapping my head around. But it looks like we:

- Add `pendingChatMembers` optimistically (it is an array of objects with accountID + the pending action)
- On failure we will overwrite the `pendingChatMembers` with the previous value of `pendingChatMembers` (why?)
- On success we will remove the `pendingChatMembers` and set the actual members

For something that will basically never fail. I do wonder why we do all this. The user would still get no feedback that anything went wrong from the look of things.

Anyways, I see we implemented it [here](https://github.com/Expensify/App/issues/34671). So, I _think_ we should do the same here for consistency, but not intent on making it a blocker.

Reply:
Following up [here](https://github.com/Expensify/App/issues/39985)